yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnShuffleEncryptionSuite.scala

```diff
@@ -0,0 +1,307 @@
+  private[this] var tempDir: File = _
+  private[this] var mergedOutputFile: File = _
+  private[this] var partitionSizesInMergedFile: Array[Long] = _
+  private[this] val ugi = UserGroupInformation.createUserForTesting("testuser", Array[String] {
+    "testgroup"
+  })
+
+  // Create a mocked shuffle handle to pass into HashShuffleReader.
+  private[this] val shuffleHandle = {
+    val dependency = mock(classOf[ShuffleDependency[Int, Int, Int]])
+    when(dependency.serializer).thenReturn(Some(serializer))
+    when(dependency.aggregator).thenReturn(None)
+    when(dependency.keyOrdering).thenReturn(None)
+    new BaseShuffleHandle(SHUFFLE_ID, NUM_MAPS, dependency)
+  }
+
+  // Make a mocked MapOutputTracker for the shuffle reader to use to determine what
+  // shuffle data to read.
+  private[this] val mapOutputTracker = mock(classOf[MapOutputTracker])
+
+  override def beforeAll(): Unit = {
+    env = new AkkaRpcEnvFactory().create(
+      RpcEnvConfig(conf, "local", "localhost", 0, new SecurityManager(conf), false))
+
+    val sparkEnv = mock(classOf[SparkEnv])
+    when(sparkEnv.rpcEnv).thenReturn(env)
+    when(sparkEnv.conf).thenReturn(conf)
+    SparkEnv.set(sparkEnv)
+
+    System.setProperty("SPARK_YARN_MODE", "true")
+    ugi.doAs(new PrivilegedExceptionAction[Unit]() {
+      override def run(): Unit = {
+        conf.set(SPARK_SHUFFLE_ENCRYPTION_ENABLED, true.toString)
+        val creds = new Credentials()
+        CryptoConf.initSparkShuffleCredentials(conf, creds)
+        SparkHadoopUtil.get.addCurrentUserCredentials(creds)
+      }
+    })
+  }
+
+  override def afterAll(): Unit = {
+    if (env != null) {
+      env.shutdown()
+    }
+    SparkEnv.set(null)
+  }
+
+  override def beforeEach(): Unit = {
+    super.beforeEach()
+    MockitoAnnotations.initMocks(this)
+    tempDir = Utils.createTempDir()
+    mergedOutputFile = File.createTempFile("mergedoutput", "", tempDir)
+  }
+
+  override def afterEach(): Unit = {
+    super.afterEach()
+    conf.set("spark.shuffle.compress", false.toString)
+    Utils.deleteRecursively(tempDir)
+    val leakedMemory = taskMemoryManager.cleanUpAllAllocatedMemory()
+    if (leakedMemory != 0) {
+      fail("Test leaked " + leakedMemory + " bytes of managed memory")
+    }
+  }
+
+  test("yarn shuffle encryption read and write") {
+    ugi.doAs(new PrivilegedExceptionAction[Unit] {
+      override def run(): Unit = {
+        testYarnShuffleEncryptionWriteRead()
+      }
+    })
+  }
+
+  test("yarn shuffle encryption read and write with shuffle compression enabled") {
+    ugi.doAs(new PrivilegedExceptionAction[Unit] {
+      override def run(): Unit = {
+        conf.set("spark.shuffle.compress", true.toString)
+        testYarnShuffleEncryptionWriteRead()
+      }
+    })
+  }
+
+  private[this] def testYarnShuffleEncryptionWriteRead(): Unit = {
+    val dataToWrite = new JArrayList[Product2[Int, Int]]()
+    for (i <- 0 to NUM_PARTITITONS) dataToWrite.add((i, i))
```

Comment:
nit:

    for (i <- 0 to NUM_PARTITITONS) {
      dataToWrite.add((i, i))
    }